### DISCUSSION TITLE
request for command line utulities for tcp tunneling

### DISCUSSION THREAD

**Discussion by phr34k** [rustdesk/rustdesk#2840] [General]
I know that this might be out of scope, but I think it'd be a fantastic addition if rustdesk could accomodate some command line utulities to initiate temp. tunnels based on id/password without having to go through rustdesk gui.  

Most of the time I prefer to manage machines through command-line e.g. ssh or powershell for convience  (e.g. automatisation) and with one-off complicated tasks where I'd actually like to interact with the desktop environment. These machines often don't have an public address (i.e. nat) making them diffifult to reach using traditional methods. Existing tools like tailscale, nebula (slack), ngrok, cloudfare cater to a different audiance and I haven't found them to be good fits.

What I propose is something akin to this: `rustdesk --tunnel 2222 192:168.1.10:22 --id xxxyyyzzz --password ./password cmd args` to let rustdesk launch an empheral tunnel the same way as rustdesk already provides, but since there's no ui involved it'd allow me to integrate it into ssh through ProxyCommand.

```
Host xxxyyyzzz
    ProxyCommand rustdesk --tunnel 2222 127.0.0.1:22 --id %h --password ./%h.key ssh localhost -p 2222
```

So instead of a manual process of opening rustdesk, connecting to the machine (re-)configurering the tcp tunnels, invoke some other commands in a terminal for the ssh session, you could streamline the entire process with a single command. Of course it isn't limited to ssh but every time I type `ssh xxxyyyzzz`  it'd be magic.

**Comment by rustdesk** [comment]
`cargo build --features cli` do what you want, but I have no plan to support it officially (no doc either), it is my test tool only.

**Comment by phr34k** [comment]
Had a brief look, so the code the cli (https://github.com/rustdesk/rustdesk/blob/master/src/cli.rs) pretty much just makes an one-off tcp tunnel? I can't really find how the entire command line works i.e. the arguments, but could you consider supporting the feature officially? I really think these are excellent features for *power* users.

**Reply by rustdesk** [reply]
Maybe in the future, you can see how many issues we have already.

**Comment by stephannn** [comment]
and if we already talking about a PS remote session, access to the local drive would also be sweet ;)

Besides I really can imagine that you have a lot of work and issues, but for the moment, an in special for the price, it looks great.

However, to explain the requirement. If you use an Azure AD Device, which goes more into direction zero-trust, you cannot work anymore with AAD Groups for local access like building up a tunnel, or check log files on the device, even in a trusted domain. To do so, every admin user has to be a direct member of a local group. This will result into unlovely local groups full of users.
Allowing admins to create a tunnel, unaware of the user, enables to check and fix devices, without disturbing them.

**Comment by phr34k** [comment]
I recently saw in some issue reports about undocumented commands lines i.e. --connect id, and when I did a search in the repository I also noticed that --file-transfer id  and --port-forward subcommands. Unfortunately, the --port-forward command isn't fully working yet, however it does seem to be what I've requested in this discussion thread.

It seems to at least start the session with tcp tunnels but it lists the old configured tunnels instead of overriding them. Would it be possible to support these overrides? 

Outside of that it also seems that the tcp tunneling itself e.g. https://github.com/rustdesk/rustdesk/issues/8702 doesn't play well anymore with ssh
